core/src/main/java/org/infinispan/interceptors/impl/ClusteringInterceptor.java

```diff
@@ -21,6 +37,47 @@
    protected DataContainer dataContainer;
    protected StateTransferManager stateTransferManager;
 
+   protected static Response getSingleReadResponseOrFail(Map<Address, Response> responseMap, CompletableFuture<?> future) {
+      Iterator<Response> it = responseMap.values().iterator();
+      if (!it.hasNext()) {
+         future.completeExceptionally(new RpcException("Expected one response"));
+         return null;
+      }
+      Response response = it.next();
+      if (it.hasNext()) {
+         future.completeExceptionally(new IllegalStateException("Too many responses " + responseMap));
+         return null;
+      }
+      if (response.isSuccessful()) {
+         return response;
+      } else if (response instanceof UnsureResponse){
+         future.completeExceptionally(OutdatedTopologyException.INSTANCE);
+      }
+      // We won't throw AllOwnersLostException for multicommands, the caller has to limit the actual keys
+      return null;
+   }
+
+   protected static Response getSingleSuccessfulResponseOrFail(Map<Address, Response> responseMap, CompletableFuture<?> future) {
```

Comment:
The difference is that getSingleReadResponseOrFail throws OTE only for UnsureResponses, while getSingleSuccessfulResponseOrFail will throw that for any unsuccessful response. But I can probably move the checks to the two uses, removing the latter.